File: 08/FunctionCalls/FibonacciElement/VMtranslator.py
__author__ = 'Or Keren'

import sys
import os
from VMparser import *
from VMCodeWriter import *

parsers = []


def main():
    """
    The main function, calls writeCode with the output path for the given file. It handles both a file argument
    and a directory argument

    **IMPORTANT**: The function does not handle incorrect or invalid files\folders
    (such as a directory that does not exist)
    :return: N/A
    """
    fpath = os.path.abspath(sys.argv[1].replace('\n', '').replace('\r', ''))
    if os.path.isfile(fpath):
        parsers.append(Parser(fpath))
        output = os.path.splitext(fpath)[0] + ".asm"

    else:
        fpath += ('/' if not str(fpath).endswith('/') else '')
        output = fpath + "/" + os.path.basename(os.path.normpath(fpath)) + '.asm'
        for file in os.listdir(fpath):
            if file.endswith(".vm") and not file.endswith("Sys.vm"):
                parsers.append(Parser(fpath + '/' + file))
        parsers.append(Parser(fpath + '/' + "Sys.vm"))

    writeCode(output)


def writeCode(output_path):
    """
    :param output_path: The output path that the writer should write the asm files into
    :return: N/A
    """
    writer = Writer(output_path)
    for parser in parsers:
        curr_path = parser.get_file_name()
        file_name = os.path.split(curr_path)[1][:-3]
        # writer = Writer(output_path + "" + file_name + ".asm")
        writer.set_file_name(file_name)

        # writer.write_bootstrap()  # TODO validate proper location - called it within ctor of writer

        while parser.can_continue():
            parser.advance()
            command_type = parser.get_command_type()

            arg1 = parser.get_first_arg()
            arg2 = parser.get_second_arg()

            # Switch - Case on command_type
            if command_type == Command.C_ARITHMETIC:
                # writer.write_arithmetic(parser.get_curr_command()) # OR - Changed it because of whitespace problems
                writer.write_arithmetic(arg1)

            elif command_type == Command.C_PUSH:
                writer.write_push(arg1, arg2)

            elif command_type == Command.C_POP:
                writer.write_pop(arg1, arg2)

            elif command_type == Command.C_LABEL:
                writer.write_label(arg1)

            elif command_type == Command.C_IF:
                writer.write_if(arg1)

            elif command_type == Command.C_GOTO:
                writer.write_goto(arg1)

            elif command_type == Command.C_FUNCTION:
                writer.write_function(arg1, arg2)

            elif command_type == Command.C_RETURN:
                writer.write_return()

            elif command_type == Command.C_CALL:
                writer.write_call(arg1, arg2)


if __name__ == "__main__":
    main()

File: 08/VMtranslator.java
import java.io.*;
import java.nio.charset.Charset;
import java.util.Vector;

/**
 * Class that manages the translation from several or one .vm file to a single .asm file.
 */
public final class VMtranslator {
	private static final int FAILURE = -1;
	private static final String USAGE_MSG = "Usage: VMtranslator <file or folder path>";
	private static final String VM_EXTENSION = ".vm";
	private static final String ASM_EXTENSION = ".asm";
	private static final String UTF8 = "UTF-8";
	private static final char DOT = '.';
	private static final String SLASH = "\\"; // TODO: need to make sure this works on linux

	/**
	 * Class constructor. A private method since this is a kind-of 'static' class that doesn't
	 * require any instantiations.
	 */
	private VMtranslator() {}

	/**
	 * The Main method. Creates VMparser instance for each .vm file it receives as input, and a
	 * single AssemblerWriter instance to translate all .vm file/s into. 
	 * @param args String Array with a single element which is the path of a .vm file or a path to
	 *             a directory which have .vm files.
	 *             If it doesn't have any .vm files then an empty .asm file is created.
	 */
	public static void main(String[] args) {
		if (args.length != 1) {
			System.out.println(USAGE_MSG);
			System.exit(FAILURE);
		}
		String path = args[0];
		try {
			File file = new File(path);
			String asmFilePath = null;
			Vector<VMparser> parsers = new Vector<>();
			if (file.isFile()) {
				parsers.add(new VMparser(file));
				asmFilePath = file.toString().substring(0, file.toString().lastIndexOf(DOT)) +
						ASM_EXTENSION;
			}
			else if (file.isDirectory()) {
				asmFilePath = file.toString() + SLASH + file.getName() + ASM_EXTENSION;
				for (File fileEntry : file.listFiles()) {
					if (fileEntry.getName().endsWith(VM_EXTENSION)) { // TODO: Or didn't append Sys.vm file
						parsers.add(new VMparser(fileEntry));
					}
				}
				// TODO: Or appended the Sys.vm file last, in this place right here.
			}
			File asmCodeFile = new File(asmFilePath);
			FileOutputStream codeFileOutStream = new FileOutputStream(asmCodeFile);
			Writer writer = new BufferedWriter(new OutputStreamWriter(codeFileOutStream,
											   Charset.forName(UTF8)));
			AssemblerWriter asmWriter = new AssemblerWriter(writer);
			for (VMparser VMparser : parsers) {
				VMparser.removedExcessLines();
				asmWriter.setFileName(VMparser.getFileName());
				// TODO: At first they called bootstrap here, then called in the ctor of writer. we don't do that in java?
				while (!VMparser.hasFinished()) {
					switch (VMparser.getCmdType()) {
						case C_ARITHMETIC:
							asmWriter.writeArithmetic(VMparser.getCurrentCommand()); // TODO: Or changed it to VMparser.getFirstArg() because of whitespace problem
							break;
						case C_PUSH:
							asmWriter.writePush(VMparser.getFirstArg(), VMparser.getSecondArg());
							break;
						case C_POP:
							asmWriter.writePop(VMparser.getFirstArg(), VMparser.getSecondArg());
							break;
						case C_LABEL:
							asmWriter.writeLabel(VMparser.getFirstArg());
							break;
						case C_IF_GOTO:
							asmWriter.writeIf(VMparser.getFirstArg());
							break;
						case C_GOTO:
							asmWriter.writeGoTo(VMparser.getFirstArg());
							break;
						case C_FUNCTION:
							asmWriter.writeFunction();
							break;
						case C_RETURN:
							asmWriter.writeReturn();
							break;
						case C_CALL:
							asmWriter.writeCall();
							break;
						default:
							System.exit(FAILURE);
					}
					VMparser.advance();
				}
			}
			asmWriter.close();
		}
		catch (Exception e) {
			System.err.println(e.getMessage());
			System.exit(FAILURE);
		}
	}
}